`{.push nodecl.}` should but doesn't apply to importc var/let symbols, can give silently wrong results

### Example 1
```nim
{.push nodecl, header: "<copyfile.h>".}
# let COPYFILE_DATA* {.nodecl.}: cint # ok
let COPYFILE_DATA*: cint # bug:  Error: 'let' symbol requires an initialization
{.pop.}
```

### Current Output
Error: 'let' symbol requires an initialization

### Expected Output
works 

### Example 2: silently wrong results with var
```nim
when true:
  {.push nodecl, header: "<copyfile.h>".}
  # var COPYFILE_DATA* {.nodecl.}: cint # ok
  var COPYFILE_DATA*: cint # bug: initializes to 0
  {.pop.}
  doAssert COPYFILE_DATA == 8, $COPYFILE_DATA # fails
```

### Example 2
{.push nodecl.} also doesn't apply for types as shown below (even though we have `{.push nodecl.}` and even though `{.push header: ....}` should imply nodecl

this example requires osx but the problem isn't specific to osx
```nim
# t11796b:
when true:
  {.push nodecl, header: "<copyfile.h>".}
  type
    # copyfile_state_t {.nodecl.} = pointer # workaround: would work
    copyfile_state_t = pointer # bug
    copyfile_flags_t = cint
  proc copyfile_state_alloc*(): copyfile_state_t
  proc c_copyfile*(src, dst: cstring,  state: copyfile_state_t, flags: copyfile_flags_t): cint {.importc: "copyfile".}
  {.pop.}

# main.nim:
impor t11796b
if false:
  let state = copyfile_state_alloc()
  discard c_copyfile("", "",  state, 0)
```

nim r -b:cpp -d:case5 $timn_D/tests/nim/all/t11796.nim
```
/Users/timothee/git_clone/nim/timn/build/nimcache/@mt11796.nim.cpp:130:13: error: no matching function for call to 'copyfile'
                int T5_ = copyfile("", "", state__WkrHkIAbpCAcQ7Yk35nGFw, ((int) 0));
                          ^~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/copyfile.h:55:5: note: candidate function not viable: cannot convert argument of incomplete type 'void *' to
      'copyfile_state_t' (aka '_copyfile_state *') for 3rd argument
int copyfile(const char *from, const char *to, copyfile_state_t state, copyfile_flags_t flags);
    ^
```

### Example 3
the docs of nodecl should be improved a bit IMO, eg:
`Note: This will not work for the LLVM backend.` => remove this, that backend was removed (and there'd be a way around this anyways)

they should also illustrate better when it's needed, and also explain that you need 2 files to see the need for header and nodecl, eg main.nim that imports fun.nim which contains c declarations (ohterwise cgen errors may be masked).


### Additional Information
1.5.1 eef2948ec2725d69358ab647a25bfa4ac675ed91
importc let without initialization was enabled since #14258
that specific example requires osx but the general bug is the same

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -528,6 +528,15 @@ proc semIdentWithPragma(c: PContext, kind: TSymKind, n: PNode,
     else: discard
   else:
     result = semIdentVis(c, kind, n, allowed)
+    case kind
+    of skType:
+      # process pragmas later, because result.typ has not been set yet
+      discard
+    of skField: implicitPragmas(c, result, n.info, fieldPragmas)
+    of skVar:   implicitPragmas(c, result, n.info, varPragmas)
+    of skLet:   implicitPragmas(c, result, n.info, letPragmas)
+    of skConst: implicitPragmas(c, result, n.info, constPragmas)
+    else: discard
 
 proc checkForOverlap(c: PContext, t: PNode, currentEx, branchIndex: int) =
   let ex = t[branchIndex][currentEx].skipConv

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -529,9 +529,6 @@ proc semIdentWithPragma(c: PContext, kind: TSymKind, n: PNode,
   else:
     result = semIdentVis(c, kind, n, allowed)
     case kind
-    of skType:
-      # process pragmas later, because result.typ has not been set yet
-      discard
     of skField: implicitPragmas(c, result, n.info, fieldPragmas)
     of skVar:   implicitPragmas(c, result, n.info, varPragmas)
     of skLet:   implicitPragmas(c, result, n.info, letPragmas)

--- a/tests/nimdoc/m13129.nim
+++ b/tests/nimdoc/m13129.nim
@@ -4,6 +4,7 @@ when defined(cpp):
   {.push header: "<vector>".}
   type
     Vector[T] {.importcpp: "std::vector".} = object
+  {.pop.}
 elif defined(js):
   proc endsWith*(s, suffix: cstring): bool {.noSideEffect,importjs: "#.endsWith(#)".}
 elif defined(c):

--- a/tests/stdlib/tsugar.nim
+++ b/tests/stdlib/tsugar.nim
@@ -12,6 +12,23 @@ type # for capture test, ref #20679
   FooCapture = ref object
     x: int
 
+proc mainProc() =
+  block: # bug #16967
+    var s = newSeq[proc (): int](5)
+    {.push exportc.}
+    proc bar() =
+      for i in 0 ..< s.len:
+        let foo = i + 1
+        capture foo:
+          s[i] = proc(): int = foo
+    {.pop.}
+
+    bar()
+
+    for i, p in s.pairs:
+      let foo = i + 1
+      doAssert p() == foo
+
 template main() =
   block: # `=>`
     block:
@@ -85,22 +102,6 @@ template main() =
           capture i, j:
             closure2 = () => (i, j)
     doAssert closure2() == (5, 3)
-
-    block: # bug #16967
-      var s = newSeq[proc (): int](5)
-      {.push exportc.}
-      proc bar() =
-        for i in 0 ..< s.len:
-          let foo = i + 1
-          capture foo:
-            s[i] = proc(): int = foo
-      {.pop.}
-
-      bar()
-
-      for i, p in s.pairs:
-        let foo = i + 1
-        doAssert p() == foo
 
     block: # issue #20679
       # this should compile. Previously was broken as `var int` is an `nnkHiddenDeref`
@@ -298,6 +299,8 @@ template main() =
       doAssert ys == @[0, 1, 2, 3, 4, 5]
 
     test()
+
+  mainProc()
 
 when not defined(js): # TODO fixme JS VM
   static:

